File: transpiler/javatests/com/google/j2cl/readable/java/innerclassinitorder/InnerClassInitOrder.java
/*
 * Copyright 2017 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package innerclassinitorder;

/**
 * Smoke test for inner classes, copied from GWT.
 */
public class InnerClassInitOrder {
  public int number = 0;

  static class OuterRefFromSuperCtorBase {
    OuterRefFromSuperCtorBase(Object o) {
      o.hashCode();
    }
  }

  class OuterRefFromSuperCtorCall extends OuterRefFromSuperCtorBase {
    OuterRefFromSuperCtorCall() {
      super(
          new Object() {
            @Override
            public int hashCode() {
              number += 100;
              return 0;
            }
          });
    }
  }

  class OuterRefFromThisCtorCall extends OuterRefFromSuperCtorBase {
    public OuterRefFromThisCtorCall(Object object) {
      super(object);
    }

    public OuterRefFromThisCtorCall() {
      this(
          new Object() {
            @Override
            public int hashCode() {
              number += 1000;
              return 0;
            }
          });
    }
  }

  public void testOuterThisFromSuperCall() {
    new OuterRefFromSuperCtorCall();
    assert (number == 100);
  }

  public void testOuterThisFromThisCall() {
    new OuterRefFromThisCtorCall();
    assert (number == 1100);
  }

  class InnerClass {
    {
      callInner();
    }

    void callInner() {
      number += 1;
      class ReallyInnerClass {
        {
          callReallyInner();
        }

        void callReallyInner() {
          number += 10;
        }
      }
      new ReallyInnerClass();
    }
  }

  static class P1<T1> {
    class P2<T2> extends P1<T1> {
      class P3<T3> extends P2<T2> {
        P3() {
          this(1);
        }

        P3(int i) {
          P2.this.super(i);
        }
      }

      P2() {
        this(1);
      }

      P2(int i) {
        super(i);
      }
    }

    final int value;

    P1() {
      this(1);
    }

    P1(int i) {
      value = i;
    }
  }


  /**
   * Used in test {@link #testExtendsNested()}
   */
  private static class ESOuter {
    class ESInner {
      public int value;

      public ESInner() {
        value = 1;
      }

      public ESInner(int value) {
        this.value = value;
      }
    }

    public ESInner newESInner() {
      return new ESInner();
    }
  }

  private static class ESInnerSubclass extends ESOuter.ESInner {
    ESInnerSubclass(ESOuter outer) {
      outer.super();
    }

    ESInnerSubclass(int value, ESOuter outer) {
      outer.super(value);
    }
  }

  /**
   * Used in test {@link #testExtendsNestedWithGenerics()}
   */
  private static class ESWGOuter<T> {
    class ESWGInner {
      public int value;

      public ESWGInner() {
        value = 1;
      }

      public ESWGInner(int value) {
        this.value = value;
      }
    }

    public ESWGInner newESWGInner() {
      return new ESWGInner();
    }
  }

  private static class ESWGInnerSubclass extends ESWGOuter<String>.ESWGInner {
    ESWGInnerSubclass(ESWGOuter<String> outer) {
      outer.super();
    }

    ESWGInnerSubclass(int value, ESWGOuter<String> outer) {
      outer.super(value);
    }
  }

  public void testExtendsNested() {
    ESOuter o = new ESOuter();
    assert (1 == o.new ESInner().value);
    assert (2 == o.new ESInner(2).value);
    assert (1 == new ESInnerSubclass(o).value);
    assert (2 == new ESInnerSubclass(2, o).value);
  }

  /**
   * Test for Issue 7789
   */
  public void testExtendsNestedWithGenerics() {
    ESWGOuter<String> o = new ESWGOuter<String>();
    assert (1 == o.new ESWGInner().value);
    assert (2 == o.new ESWGInner(2).value);
    assert (1 == new ESWGInnerSubclass(o).value);
    assert (2 == new ESWGInnerSubclass(2, o).value);
  }

  public void testInnerClassCtors() {
    P1<?> p1 = new P1<Object>();
    assert (1 == p1.value);
    assert (2 == new P1<Object>(2).value);
    P1<?>.P2<?> p2 = p1.new P2<Object>();
    assert (1 == p2.value);
    assert (2 == p1.new P2<Object>(2).value);
    assert (1 == p2.new P3<Object>().value);
    assert (2 == p2.new P3<Object>(2).value);
  }

  public void testInnerClassInitialization() {
    new InnerClass();
    assert (number == 1111);
  }

  public void testInnerClassLoop() {
    abstract class AddNumber {
      int num;

      public AddNumber(int i) {
        this.num = i;
      }

      public abstract void act();
    }
    AddNumber[] results = new AddNumber[10];
    for (int i = 0; i < 10; i++) {
      AddNumber ap = new AddNumber(i) {
        @Override
        public void act() {
          number += num;
        }
      };
      results[i] = ap;
    }
    for (AddNumber theAp : results) {
      theAp.act();
    }
    assert (number == 1156);
  }

  public static class Outer {
    public class OuterIsNotSuper {
      public int getValue() {
        return value;
      }
    }

    public class OuterIsSuper extends Outer {
      public OuterIsSuper(int i) {
        super(i);
      }

      @Override
      public int checkDispatch() {
        return 2;
      }

      public int checkDispatchFromSub1() {
        return super.checkDispatch();
      }

      public int checkDispatchFromSub2() {
        return new Outer(1) {
          public int go() {
            return OuterIsSuper.super.checkDispatch();
          }
        }.go();
      }

      public OuterIsNotSuper unqualifiedAlloc() {
        return new OuterIsNotSuper();
      }
    }

    public static class TestQualifiedSuperCall extends OuterIsNotSuper {
      public TestQualifiedSuperCall() {
        new Outer(1).new OuterIsSuper(2).super();
      }
    }

    public class TestUnqualifiedSuperCall extends OuterIsNotSuper {
      public TestUnqualifiedSuperCall() {
        super();
      }
    }

    protected final int value;

    public Outer(int i) {
      value = i;
    }

    public int checkDispatch() {
      return 1;
    }
  }

  private final Outer outer = new Outer(1);

  private final Outer.OuterIsSuper outerIsSuper = outer.new OuterIsSuper(2);

  public void testOuterIsNotSuper() {
    Outer.OuterIsNotSuper x = outerIsSuper.new OuterIsNotSuper();
    assert (2 == x.getValue());
  }

  // new an anonymous class of an inner class with a qualifier
  public void testOuterIsNotSuperAnon() {
    Outer.OuterIsNotSuper x = outerIsSuper.new OuterIsNotSuper() {
    };
    assert (2 == x.getValue());
  }

  public void testQualifiedSuperCall() {
    Outer.TestQualifiedSuperCall x = new Outer.TestQualifiedSuperCall();
    assert (2 == x.getValue());
  }

  public void testQualifiedSuperCallAnon() {
    Outer.TestQualifiedSuperCall x = new Outer.TestQualifiedSuperCall() {};
    assert (2 == x.getValue());
  }

  public void testSuperDispatch() {
    assert (1 == outerIsSuper.checkDispatchFromSub1());
    assert (1 == outerIsSuper.checkDispatchFromSub2());
  }

  public void testUnqualifiedAlloc() {
    Outer.OuterIsNotSuper x = outerIsSuper.unqualifiedAlloc();
    assert (2 == x.getValue());
  }

  public void testUnqualifiedSuperCall() {
    Outer.TestUnqualifiedSuperCall x = outerIsSuper.new TestUnqualifiedSuperCall();
    assert (2 == x.getValue());
  }

  // new an anonymous class of an inner class with a qualifier.
  public void testUnqualifiedSuperCallAnon() {
    Outer.TestUnqualifiedSuperCall x = outerIsSuper.new TestUnqualifiedSuperCall() {
    };
    assert (2 == x.getValue());
  }

  public static void main(String... args) {
    InnerClassInitOrder m = new InnerClassInitOrder();
    m.testOuterThisFromSuperCall();
    m.testOuterThisFromThisCall();

    m.testExtendsNested();
    m.testExtendsNestedWithGenerics();
    m.testInnerClassCtors();
    m.testInnerClassInitialization();
    m.testInnerClassLoop();

    m.testOuterIsNotSuper();
    m.testOuterIsNotSuperAnon();
    m.testQualifiedSuperCall();
    m.testQualifiedSuperCallAnon();
    m.testSuperDispatch();
    m.testUnqualifiedAlloc();
    m.testUnqualifiedSuperCall();
    m.testUnqualifiedSuperCallAnon();
  }
}

File: transpiler/javatests/com/google/j2cl/readable/kotlin/innerclassinitorder/InnerClassInitOrder.kt
/*
 * Copyright 2022 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package innerclassinitorder

/** Smoke test for inner classes, copied from GWT. */
class InnerClassInitOrder {
  var number = 0

  open class OuterRefFromSuperCtorBase(o: Any) {
    init {
      o.hashCode()
    }
  }

  inner class OuterRefFromSuperCtorCall :
    OuterRefFromSuperCtorBase(
      object : Any() {
        override fun hashCode(): Int {
          number += 100
          return 0
        }
      }
    )

  inner class OuterRefFromThisCtorCall(o: Any) : OuterRefFromSuperCtorBase(o) {
    constructor() :
      this(
        object : Any() {
          override fun hashCode(): Int {
            number += 1000
            return 0
          }
        }
      )
  }

  fun testOuterThisFromSuperCall() {
    OuterRefFromSuperCtorCall()
    val a = number == 100
  }

  fun testOuterThisFromThisCall() {
    OuterRefFromThisCtorCall()
    val a = number == 1100
  }

  inner class InnerClass {
    init {
      callInner()
    }

    fun callInner() {
      number += 1
      class ReallyInnerClass {
        init {
          callReallyInner()
        }

        fun callReallyInner() {
          number += 10
        }
      }
      ReallyInnerClass()
    }
  }

  open class P1<T1>(val value: Int) {
    open inner class P2<T2>(i: Int) : P1<T1>(i) {

      inner class P3<T3>(i: Int) : P1<T1>.P2<T2>(i) {

        // TODO(b/242791935): Enable after kotlin bug is fixed
        // constructor(): this(1)

      }

      // constructor(): this(1)

    }

    constructor() : this(1)
  }

  /** Used in test [testExtendsNested] */
  private class ESOuter {
    open inner class ESInner {
      var value: Int

      constructor() {
        value = 1
      }

      constructor(value: Int) {
        this.value = value
      }
    }

    fun newESInner(): ESInner {
      return ESInner()
    }

    fun newESInner(value: Int): ESInner {
      return ESInner(value)
    }
  }

  /** Used in test [testExtendsNestedWithGenerics] */
  private class ESWGOuter<T> {
    open inner class ESWGInner {
      var value: Int

      constructor() {
        value = 1
      }

      constructor(value: Int) {
        this.value = value
      }
    }

    fun newESWGInner(): ESWGInner {
      return ESWGInner()
    }

    fun newESWGInner(value: Int): ESWGInner {
      return ESWGInner(value)
    }
  }

  fun testExtendsNested() {
    val o = ESOuter()
    val a = 1 == o.ESInner().value
    val b = 2 == o.ESInner(2).value
    val a2 = 1 == o.newESInner().value
    val b2 = 2 == o.newESInner(2).value
  }

  /** Test for Issue 7789 */
  fun testExtendsNestedWithGenerics() {
    val o = ESWGOuter<String>()
    val a = 1 == o.ESWGInner().value
    val b = 2 == o.ESWGInner(2).value
    val a2 = 1 == o.newESWGInner().value
    val b2 = 2 == o.newESWGInner(2).value
  }

  fun testInnerClassCtors() {
    val p1: P1<*> = P1<Any>()
    val a = 1 == p1.value
    val b = 2 == P1<Any>(2).value
    // P2 and P3 class is commented out above.
    // val p2: P2<*> = p1.P2<Any>()
    // val c = 1 == p2.value
    val d = 2 == p1.P2<Any>(2).value
    // val e = 1 == p2.P3<Any>().value
    // val f = 2 == p2.P3<Any>(2).value
  }

  fun testInnerClassInitialization() {
    InnerClass()
    val a = number == 1111
  }

  fun testInnerClassLoop() {
    abstract class AddNumber(var num: Int) {
      abstract fun act()
    }

    val results = arrayOfNulls<AddNumber>(10)
    for (i in 0..9) {
      val ap: AddNumber =
        object : AddNumber(i) {
          override fun act() {
            number += num
          }
        }
      results[i] = ap
    }
    for (theAp in results) {
      theAp!!.act()
    }
    val a = number == 1156
  }

  open class Outer(protected val value: Int) {
    open inner class OuterIsNotSuper {
      fun getValue(): Int {
        return value
      }
    }

    inner class OuterIsSuper(i: Int) : Outer(i) {
      override fun checkDispatch(): Int {
        return 2
      }

      fun checkDispatchFromSub1(): Int {
        return super.checkDispatch()
      }

      fun checkDispatchFromSub2(): Int {
        return object : Outer(1) {
            fun go(): Int {
              return super@OuterIsSuper.checkDispatch()
            }
          }
          .go()
      }

      fun unqualifiedAlloc(): OuterIsNotSuper {
        return OuterIsNotSuper()
      }
    }

    inner class TestUnqualifiedSuperCall : OuterIsNotSuper()

    open fun checkDispatch(): Int {
      return 1
    }
  }

  private val outer = Outer(1)

  private val outerIsSuper = outer.OuterIsSuper(2)

  fun testOuterIsNotSuper() {
    val x = outerIsSuper.OuterIsNotSuper()
    val a = 2 == x.getValue()
  }

  fun testSuperDispatch() {
    val a = 1 == outerIsSuper.checkDispatchFromSub1()
    val b = 1 == outerIsSuper.checkDispatchFromSub2()
  }

  fun testUnqualifiedAlloc() {
    val x = outerIsSuper.unqualifiedAlloc()
    val a = 2 == x.getValue()
  }

  fun testUnqualifiedSuperCall() {
    val x = outerIsSuper.TestUnqualifiedSuperCall()
    val a = 2 == x.getValue()
  }
}

fun main(args: Array<String>) {
  val m = InnerClassInitOrder()
  m.testOuterThisFromSuperCall()
  m.testOuterThisFromThisCall()

  m.testExtendsNested()
  m.testExtendsNestedWithGenerics()
  m.testInnerClassCtors()
  m.testInnerClassInitialization()
  m.testInnerClassLoop()

  m.testOuterIsNotSuper()
  m.testSuperDispatch()
  m.testUnqualifiedAlloc()
  m.testUnqualifiedSuperCall()
}
